AlphaWallet/Core/Types/Tokens/TokensProcessingPipeline.swift

```diff
@@ -0,0 +1,234 @@
+        self.assetDefinitionStore = assetDefinitionStore
+    }
+
+    func tokens(for servers: [RPCServer]) -> [Token] {
+        tokensService.tokens(for: servers)
+    }
+
+    func mark(token: TokenIdentifiable, isHidden: Bool) {
+        tokensService.mark(token: token, isHidden: isHidden)
+    }
+
+    func refresh() {
+        tokensService.refresh()
+    }
+    
+    func start() {
+        tokensService.start()
+        startTickersHandling()
+        preparePipeline()
+    }
+
+    deinit {
+        print("XXX.\(self).deinit for wallet: \(wallet)")
+    }
+
+    func refreshBalance(updatePolicy: TokenBalanceFetcher.RefreshBalancePolicy) {
+        tokensService.refreshBalance(updatePolicy: updatePolicy)
+    }
+
+    func tokenViewModel(for contract: AlphaWallet.Address, server: RPCServer) -> TokenViewModel? {
+        return tokensService.token(for: contract, server: server)
+            .flatMap { TokenViewModel(token: $0) }
+            .flatMap { [weak self] in self?.applyTicker(token: $0) }
+            .flatMap { [weak self] in self?.applyTokenScriptOverrides(token: $0) }
+    }
+
+    func tokenViewModelPublisher(for contract: AlphaWallet.Address, server: RPCServer) -> AnyPublisher<TokenViewModel?, Never> {
+        let whenTickersHasChanged: AnyPublisher<Token?, Never> = coinTickersFetcher.tickersDidUpdate.dropFirst()
+            .map { [tokensService] _ in tokensService.token(for: contract, server: server) }
+            .receive(on: RunLoop.main)
+            .eraseToAnyPublisher()
+
+        let whenSignatureOrBodyChanged: AnyPublisher<Token?, Never> = assetDefinitionStore
+            .assetsSignatureOrBodyChange(for: contract)
+            .map { [tokensService] _ in tokensService.token(for: contract, server: server) }
+            .receive(on: RunLoop.main)
+            .eraseToAnyPublisher()
+
+        let whenEventHasChanged: AnyPublisher<Token?, Never> = eventsDataStore
+            .recentEventsChangeset(for: contract)
+            .filter({ changeset in
+                switch changeset {
+                case .update(let events, _, let insertions, let modifications):
+                    return !insertions.map { events[$0] }.isEmpty || !modifications.map { events[$0] }.isEmpty
+                case .initial, .error:
+                    return false
+                }
+            }).map { [tokensService] _ in tokensService.token(for: contract, server: server) }
+            .receive(on: RunLoop.main)
+            .eraseToAnyPublisher()
+
+        return Publishers.Merge4(tokensService.tokenPublisher(for: contract, server: server), whenTickersHasChanged, whenSignatureOrBodyChanged, whenEventHasChanged)
+            .map { $0.flatMap { TokenViewModel(token: $0) } }
+            .map { [weak self] in self?.applyTicker(token: $0) }
+            .map { [weak self] in self?.applyTokenScriptOverrides(token: $0) }
+            .removeDuplicates(by: { return $0?.hashValue == $1?.hashValue })
```

Comment:
Why do we need to call and compare `.hasValue` explicitly? Is it because we override `==` to compare contract addresses only, but we want to check the balance etc here?

Reply:
this is needed especially when we resolve coin tickers, result tokenViewModelPublisher might be triggered even if token related ticker id hasn't changed, so wat to avoid such events, the same for tokens collection events there might be too many events, even if we apply diffable data source, sorting, filtering, is going to be applied to snapshot even collection elements haven't changed.

Reply:
yes, as we override == operator, we might need make full comparison, for all fields